crates/ide_completion/src/render/compound.rs

```diff
@@ -0,0 +1,93 @@
+//! Code common to structs, unions, and enum variants.
+
+use crate::render::RenderContext;
+use hir::{db::HirDatabase, HasAttrs, HasVisibility, HirDisplay};
+use ide_db::SnippetCap;
+use itertools::Itertools;
+
+/// A rendered struct, union, or enum variant, split into fields for actual
+/// auto-completion (`literal`, using `field: ()`) and display in the
+/// completions menu (`detail`, using `field: type`).
+pub(crate) struct RenderedCompound {
+    pub(crate) literal: String,
+    pub(crate) detail: String,
+}
+
+/// Render a record type (or sub-type) to a `RenderedCompound`. Use `None` for
+/// the `name` argument for an anonymous type.
+pub(crate) fn render_record(
+    db: &dyn HirDatabase,
+    snippet_cap: Option<SnippetCap>,
+    fields: &[hir::Field],
+    name: Option<&str>,
+) -> RenderedCompound {
+    let fields = fields.iter();
+
+    let (completions, types): (Vec<_>, Vec<_>) = fields
+        .enumerate()
+        .map(|(idx, field)| {
+            (
+                if snippet_cap.is_some() {
+                    format!("{}: ${{{}:()}}", field.name(db), idx + 1)
+                } else {
+                    format!("{}: ()", field.name(db))
+                },
+                format!("{}: {}", field.name(db), field.ty(db).display(db)),
+            )
+        })
+        .unzip();
+    RenderedCompound {
+        literal: format!("{} {{ {} }}", name.unwrap_or(""), completions.iter().format(", ")),
+        detail: format!("{} {{ {} }}", name.unwrap_or(""), types.iter().format(", ")),
+    }
+}
+
+/// Render a tuple type (or sub-type) to a `RenderedCompound`. Use `None` for
+/// the `name` argument for an anonymous type.
+pub(crate) fn render_tuple(
+    db: &dyn HirDatabase,
+    snippet_cap: Option<SnippetCap>,
+    fields: &[hir::Field],
+    name: Option<&str>,
+) -> RenderedCompound {
+    let fields = fields.iter();
+
+    let (completions, types): (Vec<_>, Vec<_>) = fields
+        .enumerate()
+        .map(|(idx, field)| {
+            (
+                if snippet_cap.is_some() {
+                    format!("${{{}:()}}", (idx + 1).to_string())
+                } else {
+                    "()".to_string()
+                },
+                field.ty(db).display(db).to_string(),
+            )
+        })
+        .unzip();
```

Comment:
Let's iterate the slice twice, once for the rendered index, once for the type. That way we don't need to allocate at all.

Reply:
And here I was thinking I was clever for only iterating the slice once 😛 

Makes sense, though

Reply:
Alright, there we go. Definitely looks a lot cleaner that way as well I think